src/sqlfluff/dialects/dialect_snowflake.py

```diff
@@ -74,7 +74,17 @@
             r"[$][a-zA-Z0-9_.]*",
             CodeSegment,
         ),
-        RegexLexer("inline_dollar_sign", r"[a-zA-Z_][a-zA-Z0-9_$]*", CodeSegment),
+        RegexLexer(
+            "inline_dollar_sign", r"(?=.*\$)[a-zA-Z_][a-zA-Z0-9_$]*", CodeSegment
+        ),
```

Comment:
Can avoid the expensive lookahead with this:

```suggestion
        RegexLexer(
            "inline_dollar_sign",
            r"[a-zA-Z_][a-zA-Z0-9_$]*\$[a-zA-Z0-9_$]*",
            CodeSegment,
        ),
```

Reply:
Nice! Committed this suggestion